### PR TITLE
Add `Access.key` and `Access.key!` support for keyword lists

### PR DESCRIPTION
**Pull request by Sleepful**
Currently the `Access.key` function supports maps and structs, this PR adds the same functionality to keyword lists. 

Added tests for the `Access.key` behavior of maps and structs too to make sure everything is `:ok`

Motivated per the first few lines of [`Access.key`](https://hexdocs.pm/elixir/1.14.2/Access.html#get/3) docs: 

> [Access](https://hexdocs.pm/elixir/1.14.2/Access.html#content) supports keyword lists ([Keyword](https://hexdocs.pm/elixir/1.14.2/Keyword.html)) and maps ([Map](https://hexdocs.pm/elixir/1.14.2/Map.html)) out of the box.

# Before this PR:
```elixir
# keyword lists
get_in([one: 1, two: 2], [:one]) # :ok
get_in([one: 1, two: 2], [Access.key(:one)]) # error

# map
get_in(%{one: 1, two: 2}, [:one]) # :ok
get_in(%{one: 1, two: 2}, [Access.key(:one)]) # :ok
```

# After this PR:
```elixir
# keyword lists
get_in([one: 1, two: 2], [:one]) # :ok
get_in([one: 1, two: 2], [Access.key(:one)]) # :ok

# map
get_in(%{one: 1, two: 2}, [:one]) # :ok
get_in(%{one: 1, two: 2}, [Access.key(:one)]) # :ok
```

### CODE REVIEW

**Comment by whatyouhide** [comment]
I’m not sure we should support keyword lists with `Access.key/2`, because:

  * Key access is linear, not like maps
  * Keyword lists can have multiple key-value pairs with the same key, so the behaviour would have to be to either target the first pair or all pairs, and both are different from what we for maps/structs

Maybe we should have this discussion in the [elixir-lang-core mailing list](https://groups.google.com/g/elixir-lang-core) first?

**Comment by josevalim** [comment]
Yes, this is pretty much by design.

While the access syntax is allowed in maps via `map[key]`,
if your map is made of predefined atom keys, you should prefer
to access those atom keys with `map.key` instead of `map[key]`,
as `map.key` will raise if the key is missing (which is not
supposed to happen if the keys are predefined).

Similarly, since structs are maps and structs have predefined
keys, they only allow the `struct.key` syntax and they do not
allow the `struct[key]` access syntax. `Access.key/1` can also
be used to construct dynamic access to structs and maps.

In a nutshell, when using `put_in/2` and friends:

    put_in(struct_or_map.key, :value)
    put_in(keyword_or_map[:key], :value)

When using `put_in/3` and friends:

    put_in(struct_or_map, [Access.key!(:key)], :value)
    put_in(keyword_or_map, [:key], :value)

This covers the dual nature of maps in Elixir, as they can be
either for structured data or as a key-value store.

**Comment by Sleepful** [comment]
> `Access.key/1` can also be used to construct dynamic access to structs and maps

@josevalim should there be a similar function, even if a different one, to create dynamic access to keyword lists? 

Currently there is `Access.at` for lists and `Access.key` for maps/structs, however keyword lists are somewhere between the two and benefit from `key` access.

from the Access.key docs:

> The returned function is typically passed as an accessor to [Kernel.get_in/2](https://hexdocs.pm/elixir/1.13.4/Kernel.html#get_in/2), [Kernel.get_and_update_in/3](https://hexdocs.pm/elixir/1.13.4/Kernel.html#get_and_update_in/3), and friends.
> The returned function uses the default value if the key does not exist. This can be used to specify defaults and safely traverse missing keys

is there a reason for which keyword lists should not include similar facilities?

**Comment by josevalim** [comment]
You can pass the atom as is, no need to call any function. Please see the examples I pasted above. :)
